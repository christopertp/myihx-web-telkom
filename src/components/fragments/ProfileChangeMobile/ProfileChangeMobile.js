import React, { createContext, useContext, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import OtpRequest from '../../elements/OtpRequest';
import ChangeMobileForm from '../../forms/ProfileChangeEM';
import RecoveryCodeForm from '../../forms/ProfileRecoveryCode';
import WhiteCardPlain from '../../layouts/WhiteCardPlain';

const Context = createContext({});

export default function ProfileChangeMobile() {
  const { subpage } = useParams();
  const { mobileNumber, resetMessage } = useSelector(s => s.profile);
  const dispatch = useDispatch();
  const history = useHistory();
  const [ mobile, setMobile ] = useState(mobileNumber);
  const [ verifyType, setVerifyType ] = useState('');

  useEffect(() => {
    subpage && history.replace('/profile/change-mobile');
  }, []);

  useEffect(() => {
    dispatch(resetMessage());
  }, [subpage]);

  const contextValue = {
    mobile, setMobile,
    verifyType, setVerifyType,
  };

  return (
    <Context.Provider value={contextValue}>
      {(sp => {
        switch (sp) {
          case 'request': return <Request />;
          case 'verify': return <Verify />;
          default: return <New />;
        }
      })(subpage)}
    </Context.Provider>
  );
}

export function New() {
  const { setMobile } = useContext(Context);
  const { fetchCheckMobile } = useSelector(s => s.profile);
  const dispatch = useDispatch();
  const onSubmit = v => {
    setMobile(v.mobile);
    dispatch(fetchCheckMobile(v.mobile));
  };

  return (
    <WhiteCardPlain back="/profile" help icon="verify" size="medium" title="Change Phone Number">
      <ChangeMobileForm onSubmit={onSubmit} />
    </WhiteCardPlain>
  );
}

export function Request() {
  const { mobile, verifyType, setVerifyType } = useContext(Context);
  const { fetchSendOtpChangeMobile, isLoading, message } = useSelector(s => s.profile);
  const dispatch = useDispatch();
  const subtitle = 'Choose one of available method to receive account activation code';
  const onClick = type => () => {
    const otpType = type === 'SMS' ? 'sms' : 'wa';
    setVerifyType(type);
    dispatch(fetchSendOtpChangeMobile(mobile, otpType));
  };

  return (
    <WhiteCardPlain back="/profile/change-mobile" help icon="verify" size="medium"
      subtitle={subtitle} title="Choose Verification Method">
      <OtpRequest isLoading={isLoading.b} message={message} mobile={mobile}
        onClick={onClick} type={verifyType} />
    </WhiteCardPlain>
  );
}

export function Verify() {
  const { mobile, verifyType } = useContext(Context);
  const { fetchSendOtpChangeMobile, fetchVerifyMobile } = useSelector(s => s.profile);
  const dispatch = useDispatch();
  const subtitle = `Please enter the recovery code sent via ${verifyType} to ${mobile.slice(-4).padStart(mobile.length, 'â€¢')}`;
  const otpType = verifyType === 'SMS' ? 'sms' : 'wa';
  const onResend = () => dispatch(fetchSendOtpChangeMobile(mobile, otpType));
  const onSubmit = v => dispatch(fetchVerifyMobile(v.recoveryCode));

  return (
    <WhiteCardPlain back="/profile/change-mobile/request" help icon="verify" size="medium"
      subtitle={subtitle} title="Verify Phone Number">
      <RecoveryCodeForm onResend={onResend} onSubmit={onSubmit} />
    </WhiteCardPlain>
  );
}
