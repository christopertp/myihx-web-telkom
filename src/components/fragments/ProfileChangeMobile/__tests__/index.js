import React, { useContext } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import ProfileChangeMobile, { New, Request, Verify } from '../ProfileChangeMobile';

useSelector.mockImplementation(fn => {
  fn({});
  return {
    fetchCheckMobile: jest.fn(),
    fetchSendOtpChangeMobile: jest.fn(),
    fetchVerifyMobile: jest.fn(),
    isLoading: { b: false },
    mobileNumber: jest.fn(),
    resetMessage: jest.fn(),
  };
});

useContext.mockImplementation(() => ({ mobile: '0812345678890', verifyType: 'SMS' }));

describe('src/components/fragments/ProfileChangeMobile', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ProfileChangeMobile />);
    expect(tree).toMatchSnapshot();

    useParams.mockImplementationOnce(() => ({ subpage: 'request' }));
    const result2 = ProfileChangeMobile();
    expect(result2.props.children.type).toBe(Request);

    useParams.mockImplementationOnce(() => ({ subpage: 'verify' }));
    const result3 = ProfileChangeMobile();
    expect(result3.props.children.type).toBe(Verify);
  });

  test('New', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<New />);
    expect(tree).toMatchSnapshot();

    const setMobile = jest.fn();
    const dispatch = jest.fn();
    useContext.mockImplementationOnce(() => ({ setMobile }));
    useDispatch.mockImplementationOnce(() => dispatch );
    const result = New();
    result.props.children.props.onSubmit({ mobile: '081234567890' });
    expect(setMobile).toHaveBeenCalledWith('081234567890');
    expect(dispatch).toHaveBeenCalled();
  });

  test('Request', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Request />);
    expect(tree).toMatchSnapshot();

    const setVerifyType = jest.fn();
    const dispatch = jest.fn();
    useContext.mockImplementationOnce(() => ({ setVerifyType }));
    useDispatch.mockImplementationOnce(() => dispatch );
    const result = Request();
    result.props.children.props.onClick('SMS')();
    expect(setVerifyType).toHaveBeenCalledWith('SMS');
    result.props.children.props.onClick('Whatsapp')();
    expect(dispatch).toHaveBeenCalledTimes(2);
  });

  test('Verify', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Verify />);
    expect(tree).toMatchSnapshot();

    const dispatch = jest.fn();
    useDispatch.mockImplementationOnce(() => dispatch );
    useContext.mockImplementationOnce(() => ({ mobile: '0812345678890', verifyType: 'Whatsapp' }));
    const result = Verify();
    result.props.children.props.onResend();
    result.props.children.props.onSubmit({ recoveryCode: '1234' });
    expect(dispatch).toHaveBeenCalledTimes(2);
  });
});
