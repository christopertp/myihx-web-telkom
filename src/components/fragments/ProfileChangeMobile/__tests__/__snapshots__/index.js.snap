// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fragments/ProfileChangeMobile New 1`] = `
<WhiteCardPlain
  back="/profile"
  close=""
  help={true}
  icon="verify"
  size="medium"
  subtitle=""
  title="Change Phone Number"
>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={false}
    forceUnregisterOnUnmount={false}
    form="profileChangeEM"
    getFormState={[Function]}
    initialValues={
      Object {
        "email": "",
        "mobile": "",
      }
    }
    keepDirtyOnReinitialize={false}
    onSubmit={[Function]}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
    validate={[Function]}
  />
</WhiteCardPlain>
`;

exports[`src/components/fragments/ProfileChangeMobile Request 1`] = `
<WhiteCardPlain
  back="/profile/change-mobile"
  close=""
  help={true}
  icon="verify"
  size="medium"
  subtitle="Choose one of available method to receive account activation code"
  title="Choose Verification Method"
>
  <OtpRequest
    isLoading={false}
    message=""
    mobile="0812345678890"
    onClick={[Function]}
    type="SMS"
  />
</WhiteCardPlain>
`;

exports[`src/components/fragments/ProfileChangeMobile Verify 1`] = `
<WhiteCardPlain
  back="/profile/change-mobile/request"
  close=""
  help={true}
  icon="verify"
  size="medium"
  subtitle="Please enter the recovery code sent via SMS to •••••••••8890"
  title="Verify Phone Number"
>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={false}
    forceUnregisterOnUnmount={false}
    form="profileRecoveryCode"
    getFormState={[Function]}
    initialValues={
      Object {
        "recoveryCode": "",
      }
    }
    keepDirtyOnReinitialize={false}
    onResend={[Function]}
    onSubmit={[Function]}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
    validate={[Function]}
  />
</WhiteCardPlain>
`;

exports[`src/components/fragments/ProfileChangeMobile render 1`] = `
<Provider
  value={
    Object {
      "mobile": [MockFunction],
      "setMobile": [MockFunction],
      "setVerifyType": [MockFunction],
      "verifyType": "",
    }
  }
>
  <New />
</Provider>
`;
